### PR TITLE
Add Islamic calendar class code generator

### PR DESCRIPTION
**Pull request by KJhellico**
## Proposed change

Add script for Islamic calendar class code generation (#1124). It uses `hijri_converter` to calculate the dates of all Islamic holidays used by the package.

## Type of change

- [ ] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [x] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency upgrade (version update)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] This PR is filed against `beta` branch of the repository
- [x] This PR doesn't contain any merge conflicts and has clean commit history
- [x] The code style looks good: `make pre-commit`
- [x] All tests pass locally: `make test`, `make tox` (we strongly encourage adding tests to your code)
- [x] The related [documentation][docs] has been added/updated (check off the box for free if no updates is required)

[contributing-guidelines]: https://github.com/dr-prodigy/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/dr-prodigy/python-holidays/tree/beta/docs/source

### CODE REVIEW

**Review by arkid15r — COMMENTED** [review]
A couple of high-level comments to address:

File: scripts/calendar/islamic_generator.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+from pathlib import Path
+
+from hijri_converter import convert
+from hijri_converter.ummalqura import HIJRI_RANGE
+
+class_name = "_IslamicLunar"
+out_file_name = "islamic.py"
+
+class_template = """#  python-holidays
```
[the remark this answers is not in the document]
Comment (KJhellico) [reply scripts/calendar/islamic_generator.py]: Done.

File: scripts/calendar/islamic_generator.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+#           ryanss <ryanssdev@icloud.com> (c) 2014-2017
+#  Website: https://github.com/dr-prodigy/python-holidays
+#  License: MIT (see LICENSE file)
+
+from pathlib import Path
+
+from hijridate import convert
+from hijridate.ummalqura import HIJRI_RANGE
+
+class_name = "_IslamicLunar"
```
Comment (arkid15r) [inline scripts/calendar/islamic_generator.py:19]: These are constants, let's make them uppercased.

File: scripts/calendar/islamic_generator.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    # MY
+    (9, 1, "RAMADAN_BEGINNING"),  # Beginning of Ramadan
+    # MY
+    (9, 17, "NUZUL_AL_QURAN"),  # Nuzul Al-Quran Day
+    # MY
+    (2, 6, "HARI_HOL_JOHOR"),  # Hari Hol of Johor
+)
+
+
+def makelist():
```
Comment (arkid15r) [inline scripts/calendar/islamic_generator.py:56]:
```suggestion
def generate_data():
```

File: scripts/calendar/islamic_generator.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    h_year_min, h_year_max = (d[0] for d in HIJRI_RANGE)
+
+    dates_array = {}
```
Comment (arkid15r) [inline scripts/calendar/islamic_generator.py:59]: I'd prefer this to be less ambiguous (array vs dict):
```suggestion
    dates = {}
```

File: scripts/calendar/islamic_generator.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+{holiday_arrays}"""
+
+holiday_array_template = """    {hol_name}_DATES = {{
+{year_dates}
+    }}
+"""
+
+year_template = "        {year}: ({dates}),"
+
+islamic_holidays = (
```
Comment (arkid15r) [inline scripts/calendar/islamic_generator.py:32]: I guess the best sorting option here would be the calendar (month/day) one?

File: scripts/calendar/islamic_generator.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+#  specific date is a holiday as fast and flexible as possible.
+#
+#  Authors: dr-prodigy <dr.prodigy.github@gmail.com> (c) 2017-2023
+#           ryanss <ryanssdev@icloud.com> (c) 2014-2017
+#  Website: https://github.com/dr-prodigy/python-holidays
+#  License: MIT (see LICENSE file)
+
+from pathlib import Path
+
+from hijridate import convert
```
Comment (arkid15r) [inline scripts/calendar/islamic_generator.py:16]: Should we create something like requirements/external.txt for this script dependencies?

File: scripts/calendar/islamic_generator.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                else:
+                    dates_array[g_year][hol_name] = [g_date]
+            else:
+                dates_array[g_year] = {hol_name: [g_date]}
+
+    g_year_min = min(dates_array.keys())
+    g_year_max = max(dates_array.keys())
+
+    hol_list = sorted(d[2] for d in islamic_holidays)
+    holiday_arrays = []
```
Comment (arkid15r) [inline scripts/calendar/islamic_generator.py:76]:
```suggestion
    holiday_names = sorted(d[2] for d in islamic_holidays)
    holiday_data = []
```

**Review by arkid15r — COMMENTED** [review]
Some minor (mostly the naming) issues:

**Review by arkid15r — COMMENTED** [review]
LGTM

**Pull request merged**